### PR TITLE
2022 03 11 sim hw redo

### PR DESCRIPTION
**Pull request by zultron**
This PR supersedes #10 and #11.

This PR moves the simulated devices out of the unit tests and into usable device classes. Sim configurations can now include sim CiA 301 and 402 devices that pretend to step through their respective state machines, and changes to the device manager may be tested against actual application code without needing real hardware.

Unfortunately, this PR is very large because it required a lot of restructuring to merge sim classes into the main code.  I'm happy to entertain ideas for how to make its review easier.

### CODE REVIEW

File: hw_device_mgr/cia_402/tests/test_device.py [hunk 1]
```diff
@@ -2,29 +2,38 @@
 from ...cia_301.tests.test_device import TestCiA301Device as _TestCiA301Device
 
 
-class TestCiA402Device(BaseCiA402TestClass, _TestCiA301Device):
+class TestCiA402Device(_TestCiA301Device, BaseCiA402TestClass):
```
Comment (vatanaksoytezer) [inline hw_device_mgr/cia_402/tests/test_device.py:5]: small nit question here, any reason to use leading underscore in `_TestCiA301Device`

File: hw_device_mgr/device.py [hunk 2]
```diff
@@ -236,3 +254,128 @@ def scan_devices(cls, sim=True):
...
+        links = set()
+        for dev_cls in cls.get_model():
+            cls.dot_link(dev_cls, classes, links)
+        rank_same_node = "; ".join(
+            c.dot_str()
+            for c in classes
+            if "category" in c.__dict__ and c.category != "all"
+        )
+        rank_same_leaf = "; ".join(c.dot_str() for c in classes if c.name)
+        gv = "strict digraph Devices {\n"
```
Comment (vatanaksoytezer) [inline hw_device_mgr/device.py:270]: Nit: what does gv stand for here?

File: scripts/ecat_pcap_decode.py [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+# $0 0f00080000000000000000371208ffffffff0800080008003f0010c846000000
+#      2 1 4 4 2 1 4 2 2 4 4 2
+import os
+import sys
+import re
+import enum
+import pyshark
+from hw_device_mgr.lcec.command import LCECCommand
+
+# $0 ~/elmo.2022-02-24.ethercat_switch_on_disabled.pcapng
```
Comment (vatanaksoytezer) [inline scripts/ecat_pcap_decode.py:19]: Don't quite understand why this files is needed and where this is used.

**Review by vatanaksoytezer — APPROVED** [review]
I felt there are very nice structural changes, a lot of new sim devices. I think I understand what you are doing here and made sense me to me overall, left just two really unimportant nitpicks below.

File: hw_device_mgr/cia_402/tests/test_device.py (hunk 1 above)
Comment (zultron) [reply hw_device_mgr/cia_402/tests/test_device.py:5]: Yeah, `pytest` looks for test classes whose name begins with `Test`, and runs the tests from each such class.  When one test class inherits from another, imported, parent test class, `pytest` will see the imported test class and run the tests from that class, too.  That's annoying when you're just trying to run the child's tests alone, so the workaround is to import the parent test class with a name that doesn't match the `pytest` pattern so it doesn't try to run those tests as well.

File: hw_device_mgr/device.py (hunk 2 above)
Comment (zultron) [reply hw_device_mgr/device.py:270]: `dot` is part of the `graphviz` project, and the crazy language is usually in files called e.g. `foo.gv` or `foo.gv.txt` (such as in [this example][1]).  I'll add a commit to make this more clear.

[1]: https://graphviz.org/Gallery/directed/bazel.html

File: scripts/ecat_pcap_decode.py (hunk 3 above)
Comment (zultron) [reply scripts/ecat_pcap_decode.py:19]: Ha ha, good eye!  This is a debugging tool, not used in normal operation.

I was having trouble with a drive faulting, and was using Wireshark to capture EtherCAT datagrams for debugging.  The problem is the PDO data is encoded as binary representation of the data in the order of the PDO mappings, and Wireshark doesn't have any information about that, so it's just sent as a binary blob.

This script uses the `pyshark` Python library to inspect packets, and given information about the PDO mappings, will unpack those binary blobs to show the decoded process data.

Down the road, I'll be adding some code that can generate the EtherCAT PDO mappings automatically from an input config file and an EtherCAT bus scan.  Then this script won't even need you to tell it about PDOs through command line args!

**Pull request merged**